File: 07-security/01-authn-authz-and-idor/nodejs/idor_enumeration.js
// Layer 7 · Topic 1 — IDOR as a missing authorization check (Node.js).
//
// One command, no arguments, no deps: `node idor_enumeration.js`.
// Seeds 1,500 invoices across three interleaved tenants, logs in as alice,
// and fires the same 1,000-request enumeration at three handler variants,
// counting per variant how many 200 responses carried an owner that is NOT
// alice. That count is the leak per 1,000 requests.
//
// Node's specific hazard, called out in the README, is middleware ordering:
// `app.use(requireAuth)` on one router feels like coverage, and a route on a
// different router silently has no check. We model that directly: the
// "vulnerable" handler is authenticated (we know it's alice) but never
// authorizes the object. Prisma's findUnique({ where: { id } }) is the exact
// analogue of store.get(id) below -- a primary-key lookup with no caller.
//
// What to look for: vulnerable leaks ~667/1000; both fixes leak 0 and answer
// a wrong-owner id with 404 (not 403), so an enumerator learns nothing.

const N_INVOICES = 1500;
const CALLER = 1; // alice
const N_REQUESTS = 1000;

class InvoiceStore {
  constructor() {
    this.rows = new Map();
    for (let i = 1; i <= N_INVOICES; i++) {
      this.rows.set(i, { id: i, ownerId: ((i - 1) % 3) + 1, amount: i * 100 });
    }
    this.requestUser = null; // the RLS "session variable"
  }
  get(id) {
    // findUnique({ where: { id } }) -- no notion of a caller.
    return this.rows.get(id) ?? null;
  }
  getScoped(id, ownerId) {
    const r = this.rows.get(id);
    return r && r.ownerId === ownerId ? r : null;
  }
  setRequestUser(ownerId) { this.requestUser = ownerId; }
  getRls(id) {
    const r = this.rows.get(id);
    if (!r || this.requestUser === null) return null;
    return r.ownerId === this.requestUser ? r : null;
  }
}

const handlerVulnerable = (store, caller, id) => {
  const r = store.get(id);
  return r ? [200, r] : [404, null];
};
const handlerFixedQuery = (store, caller, id) => {
  const r = store.getScoped(id, caller);
  return r ? [200, r] : [404, null];
};
const handlerFixedRls = (store, caller, id) => {
  store.setRequestUser(caller); // SET LOCAL app.current_user
  const r = store.getRls(id);
  return r ? [200, r] : [404, null];
};

function* enumerationIds(n) {
  // Deterministic LCG: reproducible, measured numbers (not chosen ones).
  let x = 1;
  for (let i = 0; i < n; i++) {
    x = (1103515245 * x + 12345) % 0x80000000;
    yield 1 + (x % N_INVOICES);
  }
}

function run(handler, label) {
  const store = new InvoiceStore();
  let leaked = 0, own = 0, notFound = 0;
  for (const id of enumerationIds(N_REQUESTS)) {
    const [status, row] = handler(store, CALLER, id);
    if (status === 200) (row.ownerId !== CALLER ? leaked++ : own++);
    else notFound++;
  }
  const [wrongStatus] = handler(store, CALLER, 2); // id 2 is bob's
  console.log(
    `  ${label.padEnd(14)} leaked=${String(leaked).padStart(4)}/${N_REQUESTS}   ` +
    `own=${String(own).padStart(4)}  not_found=${String(notFound).padStart(4)}   ` +
    `wrong-owner id -> HTTP ${wrongStatus}`
  );
}

console.log("Layer 7 · Topic 1 — IDOR enumeration (logged in as alice, id=1)");
console.log(`seed: ${N_INVOICES} invoices, 3 tenants interleaved, alice owns ${N_INVOICES / 3 | 0}`);
console.log(`attack: ${N_REQUESTS} requests, ids uniform over 1..${N_INVOICES}\n`);
console.log("  handler        wrong-owner rows leaked per 1,000 requests");
run(handlerVulnerable, "vulnerable");
run(handlerFixedQuery, "fixed_query");
run(handlerFixedRls, "fixed_rls");
console.log("\nRead: authentication (we KNOW it's alice) is not authorization " +
  "(may alice touch THIS object). The vulnerable handler has the first and " +
  "skips the second; the fixes push the check into the fetch, where it cannot " +
  "return a row alice does not own.");

File: 07-security/01-authn-authz-and-idor/python/idor_enumeration.py
"""
Layer 7 · Topic 1 — IDOR as a missing authorization check (Python).

Runs entirely in-process: no DB, no network, one command, no arguments.
It seeds 1,500 invoices (ids 1..1500) across three tenants -- alice, bob,
carol -- INTERLEAVED so enumeration hits all three, then fires the same
1,000-request enumeration (logged in as alice) at three handler variants
and counts, per variant, how many responses were 200 with an owner that is
NOT alice. That count is the leak: rows a different logged-in user walked
off with, per 1,000 requests.

The three variants are the three enforcement layers from the README:

  vulnerable   -- store.get(id): a primary-key lookup that has no idea who
                  is asking. This is db.get(Invoice, id) / findUnique / findById.
  fixed_query  -- store.get_scoped(id, caller): WHERE id=? AND owner_id=?.
                  The wrong-owner row is NEVER fetched; a mismatch is a 404.
  fixed_rls    -- the handler still writes the naive get(), but the DATA
                  LAYER carries the caller and refuses to return rows the
                  caller does not own. This is the in-process analogue of
                  Postgres Row-Level Security: forgetting the filter in the
                  handler now fails CLOSED. (For the real Postgres version,
                  with the SET vs SET LOCAL pooling footgun, see
                  idor_rls_postgres.py.)

What to look for in the output: `vulnerable` leaks ~2/3 of requests (alice
owns 1/3 of the ids, so a uniformly random id belongs to someone else 2/3
of the time); both fixed variants leak exactly 0 and return 404 -- not 403 --
on a wrong-owner id, which is the enumeration-relevant difference.
"""
from dataclasses import dataclass

N_INVOICES = 1500
TENANTS = {1: "alice", 2: "bob", 3: "carol"}
CALLER = 1  # we are logged in as alice for the whole run
N_REQUESTS = 1000


@dataclass(frozen=True)
class Invoice:
    id: int
    owner_id: int
    amount_cents: int


class InvoiceStore:
    """Stands in for the invoices table. The three access methods are the
    three enforcement layers; a handler picks exactly one."""

    def __init__(self):
        # Interleaved ownership: id 1->alice, 2->bob, 3->carol, 4->alice...
        self._rows = {
            i: Invoice(id=i, owner_id=((i - 1) % 3) + 1, amount_cents=i * 100)
            for i in range(1, N_INVOICES + 1)
        }
        self._request_user = None  # the RLS "session variable"

    # --- vulnerable: lookup by primary key, no notion of a caller ---
    def get(self, invoice_id):
        return self._rows.get(invoice_id)

    # --- query-layer fix: the owner filter is part of the fetch ---
    def get_scoped(self, invoice_id, owner_id):
        row = self._rows.get(invoice_id)
        return row if row is not None and row.owner_id == owner_id else None

    # --- data-layer fix (RLS analogue): filter applied below the handler ---
    def set_request_user(self, owner_id):
        self._request_user = owner_id

    def get_rls(self, invoice_id):
        # Same shape as get(), but the policy is enforced here, where the
        # handler author cannot forget it.
        row = self._rows.get(invoice_id)
        if row is None or self._request_user is None:
            return None
        return row if row.owner_id == self._request_user else None


def handler_vulnerable(store, caller, invoice_id):
    row = store.get(invoice_id)                      # <-- trusts the id
    return (200, row) if row else (404, None)


def handler_fixed_query(store, caller, invoice_id):
    row = store.get_scoped(invoice_id, caller)       # WHERE id AND owner_id
    return (200, row) if row else (404, None)


def handler_fixed_rls(store, caller, invoice_id):
    store.set_request_user(caller)                   # SET LOCAL app.current_user
    row = store.get_rls(invoice_id)                  # naive get(), policy below
    return (200, row) if row else (404, None)


def enumeration_ids(n):
    """Deterministic LCG so the run is reproducible and the numbers are
    measured, not chosen. IDs drawn ~uniformly from 1..N_INVOICES."""
    x = 1
    for _ in range(n):
        x = (1103515245 * x + 12345) & 0x7FFFFFFF
        yield 1 + (x % N_INVOICES)


def run(handler, label):
    store = InvoiceStore()
    leaked = ok_own = not_found = 0
    for invoice_id in enumeration_ids(N_REQUESTS):
        status, row = handler(store, CALLER, invoice_id)
        if status == 200:
            if row.owner_id != CALLER:
                leaked += 1
            else:
                ok_own += 1
        else:
            not_found += 1
    # Status a KNOWN wrong-owner id returns (id 2 is bob's):
    wrong_status, _ = handler(store, CALLER, 2)
    print(f"  {label:<14} leaked={leaked:>4}/{N_REQUESTS}   "
          f"own={ok_own:>4}  not_found={not_found:>4}   "
          f"wrong-owner id -> HTTP {wrong_status}")
    return leaked


def main():
    print("Layer 7 · Topic 1 — IDOR enumeration (logged in as alice, id=1)")
    print(f"seed: {N_INVOICES} invoices, 3 tenants interleaved, "
          f"alice owns {N_INVOICES // 3} of them")
    print(f"attack: {N_REQUESTS} requests, ids uniform over 1..{N_INVOICES}\n")
    print("  handler        wrong-owner rows leaked per 1,000 requests")
    run(handler_vulnerable, "vulnerable")
    run(handler_fixed_query, "fixed_query")
    run(handler_fixed_rls, "fixed_rls")
    print("\nRead: vulnerable leaks ~667/1000 because a uniform id is not "
          "alice's 2/3 of the time; both fixes leak 0. The fixes return 404, "
          "not 403 -- a 403 would confirm the id EXISTS, still leaking "
          "existence to an enumerator. 404 leaks nothing.")


if __name__ == "__main__":
    main()
